[DOCS] Zeppelin Flink Spark tutorial

rawkintrevo opened the pull request:
Moon recommended I migrate a How To tutorial I wrote into the website. 

http://mail-archives.apache.org/mod_mbox/incubator-zeppelin-users/201511.mbox/browser

I didn't create a JIRA issue, it doesn't look like that is nessicary for website updates? If it is, I'm sorry I'll go make one. Sorry, I'm new.

felixcheung commented:
If I understand correctly this requires a special build of Flink?

rawkintrevo commented:
No.  Just the Flink 0.9 release that Zeppelin runs on internally.

felixcheung commented:
I see, so this is referring to 0.10 Flink only?
"
http://trevorgrant.org/2015/11/03/apache-casserole-a-delicious-big-data-recipe-for-the-whole-family/
It runs against Flink 0.10 (Till Rohrmann's branch) and Spark 1.4. It would
probably help if Till's support for Flink 0.10 was merged back in, then I
could probably also get it working against Spark 1.5 (or 1.6).  As it was,
those two (Spark 1.5 and 1.6) were what really killed me the last few days,
I finally gave up and just used 1.4 which worked well."

And this https://mail-archives.apache.org/mod_mbox/flink-user/201510.mbox/%3CCAC27z=OhB5pRSeyHodU5WvdJ4Y+biBpefh9eTvh_MYvvqgvooA@mail.gmail.com%3E

felixcheung commented on docs/docs/tutorial/installing_with_flink_and_spark_cluster.md:
````diff
@@ -0,0 +1,244 @@
+build-target/bin/start-cluster.sh
+```
+
+In a browser, navigate to http://`yourip`:8082 to see the Flink Web-UI.  Click on 'Task Managers' in the left navigation bar. Ensure there is at least one Task Manager present. [Screenshot](https://trevorgrantdotorg.files.wordpress.com/2015/10/flink-dash.png)
````
how reliable is this image link? should we host this from the doc site?

felixcheung commented on docs/docs/tutorial/installing_with_flink_and_spark_cluster.md:
````diff
@@ -0,0 +1,244 @@
+
+##### Flink Test
+Create a new notebook named "Flink Test" and copy and paste the following code.
+```
````
```scala for syntax highlighting

felixcheung commented on docs/docs/tutorial/installing_with_flink_and_spark_cluster.md:
````diff
@@ -0,0 +1,244 @@
+
+##### Spark Test
+Create a new notebook named "Spark Test" and copy and paste the following code.
+```
````
```scala here as well

rawkintrevo commented:
Ahh, yes the blog version runs against a branch of Zeppelin that covers the most up to date Flink.  That seemed inappropriate here so these directions refer to Flink 0.9.  

The image will be there for the next year anyway but, hosting on the doc site is much wiser. I was parsimonious with my images because I didn't want to be linking back to me.  I can move that over and sprinkle in a few more. 

Where is the appropriate spot? 
./images
../images
../../assets/images
?

felixcheung commented:
docs/assets/

like in: docs/assets/themes/zeppelin/img

rawkintrevo commented:
After carefully reviewing 
https://github.com/rawkintrevo/incubator-zeppelin/blob/master/docs/docs/install/yarn_install.md

I don't think my contribution adds much other than 'How To set up Zeppelin with Flink'. 
I think it would be better to scale the entire thing down to a how-to tutorial focused on that. 

Thoughts?

epugh commented:
@rawkintrevo I found it very valuable...   I'd never set up Spark or Flink, and your tutorial assumed a much lower degree of knowledge.

jongyoul commented:
@rawkintrevo The failed test is removed by 11/11/2015. If you rebase it from current master CI will pass

bzz commented:
@rawkintrevo do you need any help finishing this one?

rawkintrevo commented:
@jongyoul @bzz Sorry, I got busy with school and work stuff.  Also, weird my gmail didn't flag these as important. Anyway, the only thing left to do is to move the images over to git hub and re link. I don't know exactly how to do that but I'm sure I can find out quickly enough.  I think I also need to rebase to catch up.  I'm going to try to knock this out this afternoon.

rawkintrevo commented:
this is failing on missing dependencies

bzz commented:
@rawkintrevo looks like [CI fails](https://s3.amazonaws.com/archive.travis-ci.org/jobs/97787215/log.txt) on RAT check for licence. 
Does all the files that you have contributed have a comment with appropriate Apache licence header? 

```
[INFO] --- apache-rat-plugin:0.11:check (verify.rat) @ zeppelin ---
...
[INFO] 138 resources included (use -debug for more details)
[INFO] Rat check: Summary of files. Unapproved: 1 unknown: 1 generated: 0 approved: 56 licence.
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] Zeppelin ........................................... FAILURE [ 11.330 s]
[INFO] Zeppelin: Interpreter .............................. SKIPPED
...
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
```

rawkintrevo commented:
I only added pages to the website, and didn't realize those need headers. Updating now.

Leemoonsoo commented:
@rawkintrevo You might also want to add link to navigation menu
https://github.com/apache/incubator-zeppelin/blob/master/docs/_includes/themes/zeppelin/_navigation.html#L30

rawkintrevo commented:
good idea- I am out for the holiday- I'll hit this Sunday night.

bzz commented:
Looks great to me!
@Leemoonsoo @rawkintrevo shall we merge this one now?

Leemoonsoo commented:
LGTM. Let's merge

rawkintrevo commented:
Yea, this is my first merge so once again- I have no idea what I'm doing. Let me know if I need to do something. 

Thanks!

bzz commented:
@rawkintrevo before merge - can you please run this doc though the spell checker one more time?

bzz commented on docs/tutorial/installing_with_flink_and_spark_cluster.md:
```diff
@@ -0,0 +1,328 @@
+
+**Note:** On build failuers. Having installed Zeppelin close to 30 times now, I will tell you that sometimes the build fails for seemingly no reason. 
+As long as you didn't edit any code, it is unlikely the build is failing because of something you did. What does tend to happen, is some dependency that maven is trying to download
+is unreachable.  If you get a build failure, just retry (that is, run the `mvn clean package -DskipTests -Pspark-1.5` again. If it keeps failing, try again tomorrow or after the weekend 
```
I think it's more productive to be specific on build failures here.

I.e we could let people know that:
 1. build failure may happen
 2. to understand the reason - they need to look through maven logs (networking issues on access to remote maven repo is one possible reason)
 3. in case that happens and reason is not clear - they can ask help on community mailing lists, file JIRA with logs attached, etc

rawkintrevo commented:
Spell check yes.

On note 2: I agree in spirit, however I'd really like to keep accessible very novice users. Going forward I'm going to be maintaining this article, not my original blog post.  One of the big (unsung) advantages of Zeppelin is it makes big data products friendly to work with (e.g. copy and paste code, no command line, no setting up an overwhelming IDE like jetbrains, etc.) Things we tend to take for granted as developers.  In my mind, for many people using this tutorial, this is their first experience building a package from source.  

My point being: Waiving my hands and saying sometimes things don't work and its not your fault, then leaving it at that is probably a little too dismissive, I want to let users know, "just because the build is failing- don't get discouraged, try it again and if it still doesn't work, reach out to the community for help, and in some cases, you just need to walk away from it for a little while, that's life."  

A blurb in there about looking through the logs is a good idea.  I'm going to disconnect my internet and try building again to get the specific error message because in my experience (having coached a few novices through the process), that really is the most likely reason for a failure (assuming you have followed the directions carefully).

chiwanpark commented on docs/tutorial/installing_with_flink_and_spark_cluster.md:
````diff
@@ -0,0 +1,328 @@
+        This creates the count variable which is a list of tuples of the form (word, occurances)
+*/
+val counts = text.flatMap{ _.toLowerCase.split("\\W+") }.map { (_,1) }.groupBy(0).sum(1)
+```
````
Could you add a `counts.print()` statement to print output of Flink?

rawkintrevo commented:
That was sloppy copy-paste from my gist. My bad...

It's supposed to be
counts.collect().foreach(println(_))  // execute the script and print each element in the counts list
Just like spark

corneadoug commented on docs/tutorial/installing_with_flink_and_spark_cluster.md:
```diff
@@ -0,0 +1,328 @@
+`-DskipTests` skips build tests- you're not developing (yet), so you don't need to do tests, the clone version *should* build.
+
+`-Pspark-1.5` tells maven to build a Zeppelin with Spark 1.5.  This is important because Zeppelin has its own Spark interpretter and the versions must ~~jive~~ be the same. 
+**Note:** You can build against any version of Spark that has a Zeppelin build profile available. The key is to make sure you check out the matching version of Spark to build. At the time of this writing, Spark 1.5 was the most recent Spark version available.
```
You might want to add a new line before the `**Note**` otherwise it's stuck on the side of previous pragraph

corneadoug commented on docs/tutorial/installing_with_flink_and_spark_cluster.md:
```diff
@@ -0,0 +1,328 @@
+`-Pspark-1.5` tells maven to build a Zeppelin with Spark 1.5.  This is important because Zeppelin has its own Spark interpretter and the versions must ~~jive~~ be the same. 
+**Note:** You can build against any version of Spark that has a Zeppelin build profile available. The key is to make sure you check out the matching version of Spark to build. At the time of this writing, Spark 1.5 was the most recent Spark version available.
+
+**Note:** On build failuers. Having installed Zeppelin close to 30 times now, I will tell you that sometimes the build fails for seemingly no reason. 
```
A typo here, `failuers` -> `failures`

rawkintrevo commented:
Bump? These edits were taken care of some time ago.

felixcheung commented on docs/tutorial/installing_with_flink_and_spark_cluster.md:
```diff
@@ -0,0 +1,339 @@
+
+Open a web-browser on a machine connected to the same network as the host (or in the host operating system if using a virtual machine).  Navigate to http://`yourip`:8080, where yourip is the IP address you found in `ifconfig`.
+
+See the [Zeppelin tutorial](../tutorial/tutorial.md) for basic Zeppelin usage. It is also advised that you take a moment to check out the tutorial notebook that is included with each Zeppelin install, and to familiarize yourself with basic notebook functionality.
```
this should be `../tutorial/tutorial.html`?

felixcheung commented on docs/tutorial/installing_with_flink_and_spark_cluster.md:
```diff
@@ -0,0 +1,339 @@
+
+##### Flink Cluster
+
+See the [Flink Installation guide](https://github.com/apache/flink/blob/master/README.md) for more detailed instructions.
```
would it be better to point to https://ci.apache.org/projects/flink/flink-docs-release-0.10/quickstart/setup_quickstart.html?

felixcheung commented on docs/tutorial/installing_with_flink_and_spark_cluster.md:
```diff
@@ -0,0 +1,339 @@
+
+##### Spark 1.5 Cluster
+
+See the [Spark Installation](https://github.com/apache/spark/blob/master/README.md) guide for more detailed instructions.
```
similarly http://spark.apache.org/docs/latest/index.html

felixcheung commented on docs/tutorial/installing_with_flink_and_spark_cluster.md:
```diff
@@ -0,0 +1,339 @@
+
+### Next Steps
+
+Check out the [tutorial](./tutorial.md) for more cool things you can do with your new toy!
```
`tutorial.html`?

AhyoungRyu replied:
`./tutorial.md` -> `../quickstart/tutorial.html`

felixcheung commented:
@bzz would you have any more thoughts on build failure described here?

felixcheung commented:
Where are we on this PR?

rawkintrevo commented:
I'm basing my tutorial at ApacheCon loosely off of this.  Let me go back through and edit/update as appropriate. Hopefully next week or so.

bzz commented:
Shall we wrap this work up?

AFAIK we just need to:
 - rebased on the latest master
 - address @felixcheung feedback
 - merge the guy

rawkintrevo commented:
@bzz so much has changed since I wrote this I want to go through it once and make sure everything is still accurate.  I'm f/t trying to get [ZEPPELIN-116](https://github.com/apache/zeppelin/pull/928) done. Will do this as soon as that is complete.

rawkintrevo commented:
@bzz took advantage of the break in the action on 116 to finish this off. Should be GTG now. 

Updated to Flink 1.0, Maven 3.3.9, Spark 1.6.  Also added some points on just using Spark/Flink binaries, as I have run into multiple headaches when trying to build all projects from source.

AhyoungRyu commented on docs/_includes/themes/zeppelin/_navigation.html:
```diff
@@ -27,6 +27,11 @@
                 <li role="separator" class="divider"></li>
                 <li class="title"><span><b>Basic Feature Guide</b><span></li>
+                <!-- li><span><b>Tutorial</b><span></li -->
+                <li><a href="{{BASE_PATH}}/tutorial/tutorial.html">Tutorial</a></li>
```
Is there any particular reason to add `tutorial.html` again? As you can see in the above, already we have 
```
<li><a href="{{BASE_PATH}}/quickstart/tutorial.html">Tutorial</a></li>
```

When I build this in my local, the dropdown menu is shown like the below.
<img width="554" alt="screen shot 2016-07-16 at 4 23 15 pm" src="https://cloud.githubusercontent.com/assets/10060731/16893406/a3a3f65e-4b71-11e6-940f-5b8050eb4037.png">

AhyoungRyu commented on docs/_includes/themes/zeppelin/_navigation.html:
```diff
@@ -27,6 +27,11 @@
+                <!-- li><span><b>Tutorial</b><span></li -->
+                <li><a href="{{BASE_PATH}}/tutorial/tutorial.html">Tutorial</a></li>
+                <li><a hrev="{{BASE_PATH}}/tutorial/installing_with_flink_and_spark_cluster.html">Install Zeppelin with Flink and Spark Clusters Tutorial</a></li>
+                <li role="separator" class="divider"></li>
```
I think we don't need another divider in here :)

AhyoungRyu commented on docs/_includes/themes/zeppelin/_navigation.html:
```diff
@@ -27,6 +27,11 @@
                 <li class="title"><span><b>Basic Feature Guide</b><span></li>
+                <!-- li><span><b>Tutorial</b><span></li -->
+                <li><a href="{{BASE_PATH}}/tutorial/tutorial.html">Tutorial</a></li>
+                <li><a hrev="{{BASE_PATH}}/tutorial/installing_with_flink_and_spark_cluster.html">Install Zeppelin with Flink and Spark Clusters Tutorial</a></li>
```
How about put this docs below [this line](https://github.com/apache/zeppelin/blob/master/docs/_includes/themes/zeppelin/_navigation.html#L35)?

rawkintrevo commented:
Good  eye @AhyoungRyu , the navigation made more sense in December, I didn't look at it again since the rebase. 

Thanks!

AhyoungRyu commented:
@rawkintrevo Yeah thanks! But still the docs existed in `Getting Started` section. Can we move this docs below [this line](https://github.com/apache/zeppelin/blob/master/docs/_includes/themes/zeppelin/_navigation.html#L35)? :)

rawkintrevo commented:
Yea, I had intentionally left in 'getting started' as it was an install tutorial, but I just moved it to the 'more section' per your comment.

rawkintrevo commented:
@bzz or whoever is in charge of this, failed on flakey test. Can we do a merge on this?

bzz commented on docs/quickstart/install_with_flink_and_spark_cluster.md:
```diff
@@ -0,0 +1,413 @@
+
+Check out the [tutorial](./tutorial.md) for more cool things you can do with your new toy!
+
+[Join the community](http://zeppelin.incubator.apache.org/community.html), ask questions and contribute! Every little bit helps.
```
I think we need to make sure that links\descriptions here and above do not include `incubator` as Apache Zeppelin has graduated.

corneadoug commented:
@rawkintrevo Lets remove the `incubator` from URLs, rebase, then merge it

rawkintrevo commented:
@corneadoug done and done.

corneadoug commented:
@rawkintrevo Did the rebase go wrong somehow? Seems there is a lot of commits that are not related to this PR

rawkintrevo commented:
Yea I think I merged it or something. I fixed.

corneadoug commented:
Awesome, Merging it
